rocAL/source/loaders/audio/audio_loader.cpp

```diff
@@ -0,0 +1,282 @@
+}
+
+void AudioLoader::set_prefetch_queue_depth(size_t prefetch_queue_depth) {
+    if (prefetch_queue_depth <= 0)
+        THROW("Prefetch queue depth value cannot be zero or negative");
+    _prefetch_queue_depth = prefetch_queue_depth;
+}
+
+void AudioLoader::set_gpu_device_id(int device_id) {
+    if (device_id < 0)
+        THROW("invalid device_id passed to loader");
+    _device_id = device_id;
+}
+
+size_t
+AudioLoader::remaining_count() {
+    return _remaining_audio_count;
+}
+
+void AudioLoader::reset() {
+    // stop the writer thread and empty the internal circular buffer
+    _internal_thread_running = false;
+    _circ_buff.unblock_writer();
+    if (_load_thread.joinable())
+        _load_thread.join();
+    // Emptying the internal circular buffer
+    _circ_buff.reset();
+    // resetting the reader thread to the start of the media
+    _audio_counter = 0;
+    _audio_loader->Reset();
+    // Start loading (writer thread) again
+    start_loading();
+}
+
+void AudioLoader::de_init() {
+    // Set running to 0 and wait for the internal thread to join
+    stop_internal_thread();
+    _output_mem_size = 0;
+    _batch_size = 1;
+    _is_initialized = false;
+}
+
+LoaderModuleStatus
+AudioLoader::load_next() {
+    return update_output_audio();
+}
+
+void AudioLoader::set_output(Tensor* output_tensor) {
+    _output_tensor = output_tensor;
+    _output_mem_size = _output_tensor->info().data_size();
+}
+
+void AudioLoader::stop_internal_thread() {
+    _internal_thread_running = false;
+    _stopped = true;
+    _circ_buff.unblock_reader();
+    _circ_buff.unblock_writer();
+    _circ_buff.reset();
+    if (_load_thread.joinable())
+        _load_thread.join();
+}
+
+void AudioLoader::initialize(ReaderConfig reader_cfg, DecoderConfig decoder_cfg, RocalMemType mem_type, unsigned batch_size, bool decoder_keep_original) {
+    if (_is_initialized)
+        WRN("initialize() function is already called and loader module is initialized")
+    if (_output_mem_size == 0)
+        THROW("output audio size is 0, set_output() should be called before initialize for loader modules")
+    _mem_type = mem_type;
+    _batch_size = batch_size;
+    _loop = reader_cfg.loop();
+    _audio_loader = std::make_shared<AudioReadAndDecode>();
+    size_t shard_count = reader_cfg.get_shard_count();
+    int device_id = reader_cfg.get_shard_id();
+    try {
+        // set the device_id for decoder same as shard_id for number of shards > 1
+        if (shard_count > 1) {
+            _audio_loader->Create(reader_cfg, decoder_cfg, _batch_size, device_id);
+        } else {
+            _audio_loader->Create(reader_cfg, decoder_cfg, _batch_size);
+        }
+    } catch (const std::exception& e) {
+        de_init();
+        throw;
+    }
+    _max_decoded_samples = _output_tensor->info().max_shape().at(0);
+    _max_decoded_channels = _output_tensor->info().max_shape().at(1);
+    _decoded_audio_info._data_names.resize(_batch_size);
```

Comment:
why can't we have a vector of structs instead of struct of multiple vectors

Reply:
These vectors are passed to different functions, where it accepts a vector of values like [update_tensor_roi ](https://github.com/swetha097/rocAL/blob/67cda83e168d31507ddd515be127fa770bc22b38/rocAL/source/loaders/image/image_loader.cpp#L267)and [get_file_names](https://github.com/swetha097/rocAL/blob/67cda83e168d31507ddd515be127fa770bc22b38/rocAL/source/loaders/image/image_loader.cpp#L306), hence we need the names, width and height as separate vectors